File: Bitwise/Medium/LET_no698_Partition to K Equal Sum Subsets.cpp
class Solution {
public:
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        int n = nums.size(), sum = 0;
        for (int i : nums) sum += i;
        if (sum % k) return false;
        
        int target = sum / k;
        vector<int> dp(1 << n, -1);
        dp[0] = 0;
        
        for (int mask = 0; mask < (1 << n); mask++) {
            if (dp[mask] == -1) continue;
            for (int i = 0; i < n; i++) {
                if (!(mask & 1 << i) && dp[mask] + nums[i] <= target)
                    dp[mask | 1 << i] = (dp[mask] + nums[i]) % target;
            }
        }
        
        return dp[(1 << n) - 1] == 0;
    }
};

File: Depth-First Search/Backtracking/Medium/LET_no698_Partition to K Equal Sum Subsets.py
class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        total = sum(nums)
        if total % k != 0:
            return False
        
        target = total // k
        nums.sort(reverse=True)
        bk = [0] * k
        
        return self.helper(bk, nums, target, 0)
    
    def helper(self, bk, nums, target, i):
        if i == len(nums):
            if any(s != target for s in bk):
                return False
            return True
        
        for j in range(len(bk)):
            if bk[j] + nums[i] <= target:
                bk[j] += nums[i]
                if self.helper(bk, nums, target, i + 1):
                    return True
                bk[j] -= nums[i]
        
        return False